app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/Model/Indexer/ModifySelectInProductPriceIndexFilter.php

```diff
@@ -0,0 +1,147 @@
+        $this->resourceConnection = $resourceConnection;
+        $this->defaultStockProvider = $defaultStockProvider;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+    }
+
+    /**
+     * Modify parent select. Add MSI stock table to select.
+     *
+     * @param ProductPriceIndexFilter $subject
+     * @param callable $closure
+     * @param IndexTableStructure $priceTable
+     * @param array $entityIds
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function aroundModifyPrice(
+        ProductPriceIndexFilter $subject,
+        callable $closure,
+        IndexTableStructure $priceTable,
+        array $entityIds = []
+    ): void {
+        if ($this->stockConfiguration->isShowOutOfStock()) {
+            return;
+        }
+
+        $tmpPriceTableName = $priceTable->getTableName();
+
+        foreach ($this->getWebsiteIdsFromTmpTable($tmpPriceTableName) as $websiteId) {
+            $stock = $this->stockByWebsiteIdResolver->execute($websiteId);
+            $stockId = (int)$stock->getStockId();
+            $stockTable = $this->stockIndexTableNameResolver->execute($stockId);
+
+            $connection = $this->resourceConnection->getConnection('indexer');
+            $select = $connection->select();
+            $select->from(['price_index' => $tmpPriceTableName], []);
+
+            if ($stockId === $this->defaultStockProvider->getId()) {
+                $select->joinLeft(
+                    ['stock_item' => $stockTable],
+                    'stock_item.product_id = price_index.' . $priceTable->getEntityField()
+                    . ' AND stock_item.stock_id = ' . $stockId,
+                    []
+                );
+            } else {
+                $select->joinInner(
+                    ['product_entity' => $connection->getTableName('catalog_product_entity')],
+                    'product_entity.entity_id = price_index.' . $priceTable->getEntityField(),
+                    []
+                )->joinLeft(
+                    ['stock_item' => $stockTable],
+                    'stock_item.sku = product_entity.sku',
+                    []
+                );
+            }
+
+            $select->where('stock_item.is_salable = 0 OR stock_item.is_salable IS NULL');
+            $select->where('price_index.website_id = ?', $websiteId);
+            $query = $select->deleteFromSelect('price_index');
+            $connection->query($query);
+        }
+    }
+
+    /**
+     * Get all website ids from price temporary table.
+     *
+     * @param string $tmpPriceTableName
+     * @return array
+     */
+    private function getWebsiteIdsFromTmpTable(string $tmpPriceTableName): array
+    {
+        $result = [];
+
+        $connection = $this->resourceConnection->getConnection('indexer');
+        $select = $connection->select();
+        $select->from(
+            ['price_index' => $tmpPriceTableName],
+            ['website_id']
+        )->distinct();
```

Comment:
no need to grab all the website IDs, you need to proceed with re-index just for websites to which product being reindexed are assigned to